internal/graph/memory.go

```diff
@@ -116,19 +117,54 @@ func (m *Memory) Predecessors(_ context.Context, node ocispec.Descriptor) ([]oci
 	return res, nil
 }
 
+// RemoveFromIndex removes the node from its predecessors and successors.
+func (m *Memory) RemoveFromIndex(ctx context.Context, node ocispec.Descriptor) error {
+	nodeKey := descriptor.FromOCI(node)
+	// remove the node from its successors' predecessor list
+	value, _ := m.successors.Load(nodeKey)
+	successors := value.(*sync.Map)
+	successors.Range(func(key, _ interface{}) bool {
+		value, _ = m.predecessors.Load(key)
+		predecessors := value.(*sync.Map)
+		predecessors.Delete(nodeKey)
+		return true
+	})
+	m.removeFromMemory(ctx, node)
+	return nil
+}
+
 // index indexes predecessors for each direct successor of the given node.
 // There is no data consistency issue as long as deletion is not implemented
 // for the underlying storage.
 func (m *Memory) index(ctx context.Context, node ocispec.Descriptor, successors []ocispec.Descriptor) {
+	m.indexIntoMemory(ctx, node)
 	if len(successors) == 0 {
 		return
 	}
-
 	predecessorKey := descriptor.FromOCI(node)
 	for _, successor := range successors {
 		successorKey := descriptor.FromOCI(successor)
-		value, _ := m.predecessors.LoadOrStore(successorKey, &sync.Map{})
-		predecessors := value.(*sync.Map)
-		predecessors.Store(predecessorKey, node)
+		// store in m.predecessors, memory.predecessors[successorKey].Store(node)
+		pred, _ := m.predecessors.LoadOrStore(successorKey, &sync.Map{})
+		predecessorsMap := pred.(*sync.Map)
+		predecessorsMap.Store(predecessorKey, node)
+		// store in m.successors, memory.successors[predecessorKey].Store(successor)
+		succ, _ := m.successors.Load(predecessorKey)
+		successorsMap := succ.(*sync.Map)
+		successorsMap.Store(successorKey, successor)
 	}
 }
+
+func (m *Memory) indexIntoMemory(ctx context.Context, node ocispec.Descriptor) {
+	key := descriptor.FromOCI(node)
+	m.predecessors.LoadOrStore(key, &sync.Map{})
+	m.successors.LoadOrStore(key, &sync.Map{})
+	m.indexed.LoadOrStore(key, &sync.Map{})
+}
+
+func (m *Memory) removeFromMemory(ctx context.Context, node ocispec.Descriptor) {
+	key := descriptor.FromOCI(node)
+	m.predecessors.Delete(key)
```

Comment:
We might not want to delete the key from `predecessors` when a node is deleted.

Reply:
Why not? Do we have a scenario of still needing it after the node is deleted? The information is stored in the node's predecessors' `successors` anyway.

Reply:
To get the predecessors of a deleted node.

Reply:
In what scenarios do we need to get the predecessors of a deleted node? It sounds nice to have though.

Reply:
Added the entry back. Currently we don't delete the entry in `predecessors` when doing `Delete`.